atoma-auth/src/auth.rs

```diff
@@ -496,6 +542,72 @@ impl Auth {
         Ok(())
     }
 
+    async fn get_zk_address(zk_login_signature: &str, tx_digest: &str) -> Result<String> {
+        let user_signature = UserSignature::from_base64(zk_login_signature)?;
+        if let UserSignature::ZkLogin(zk_login_authenticator) = user_signature {
+            // The message is the transaction digest that we claim is ours
+            let intent_msg = IntentMessage::new(
+                Intent::personal_message(),
+                PersonalMessage {
+                    message: tx_digest.as_bytes().to_vec(),
+                },
+            );
+
+            let intent_bcs = bcs::to_bytes(&intent_msg).unwrap();
+            let message_hash = blake2b_hash(&intent_bcs);
+            // First verify that the signature is valid
+            match zk_login_authenticator.signature {
+                SimpleSignature::Ed25519 {
+                    signature,
+                    public_key,
+                } => {
+                    let public_key = Ed25519PublicKey::from_bytes(public_key.as_ref()).unwrap();
+                    let signature = Ed25519Signature::from_bytes(signature.as_ref()).unwrap();
+                    public_key.verify(message_hash.as_slice(), &signature)?;
+                }
+                SimpleSignature::Secp256k1 {
+                    signature,
+                    public_key,
+                } => {
+                    let public_key = Secp256k1PublicKey::from_bytes(public_key.as_ref()).unwrap();
+                    let signature = Secp256k1Signature::from_bytes(signature.as_ref()).unwrap();
+                    public_key.verify(message_hash.as_slice(), &signature)?;
+                }
+                SimpleSignature::Secp256r1 {
+                    signature,
+                    public_key,
+                } => {
+                    let public_key = Secp256r1PublicKey::from_bytes(public_key.as_ref()).unwrap();
+                    let signature = Secp256r1Signature::from_bytes(signature.as_ref()).unwrap();
+                    public_key.verify(message_hash.as_slice(), &signature)?;
+                }
+            };
+            // Now that we know the signature is valid, lets get the address from the signature
+            let address_seed =
+                Bn254FrElement::from_str(&zk_login_authenticator.inputs.address_seed.to_string())
+                    .map_err(|e| {
+                    AuthError::FailedToParseSignature(format!(
+                        "Failed to parse address seed: {}",
+                        e
+                    ))
+                })?;
+            let public_id =
+                ZkLoginPublicIdentifier::new(google::ISS, &address_seed).map_err(|e| {
+                    AuthError::FailedToParseSignature(format!(
+                        "Failed to parse public identifier: {}",
+                        e
+                    ))
+                })?;
+            let public_key = PublicKey::ZkLogin(public_id);
```

Comment:
Shouldn't this public key be the same as the one that verifies the signature, above ?

Reply:
No. Private key for zklogin transaction doesn't correspond to the address. The address is generated from the zk data.

Reply:
I see